[LLDB] Run API tests with PDB too

Nerixyz opened the pull request:
From https://github.com/llvm/llvm-project/pull/148554#issuecomment-3083261858 - this adds an option for API tests to be run with the both PDB readers on Windows. As there are a lot of failures with PDB, this is an opt-in per test.

To get PDB, `-g -gcodeview` has to be used on Clang. `-gcodeview` alone isn't enough, because it won't cause clang to pass `-debug` to the linker.

#149498 tracks the (currently) failing tests.

DavidSpickett commented:
I ran this on Linaro's bot machine and got these results:
```
********************
Unresolved Tests (5):
  lldb-api :: functionalities/data-formatter/data-formatter-stl/generic/string/TestDataFormatterStdString.py
  lldb-api :: lang/c/vla/TestVLA.py
  lldb-api :: python_api/frame/get-variables/TestGetVariables.py
  lldb-api :: python_api/process/TestProcessAPI.py
  lldb-api :: tools/lldb-dap/commands/TestDAP_commands.py

********************
Failed Tests (119):
  lldb-api :: commands/dwim-print/TestDWIMPrint.py
  lldb-api :: commands/expression/anonymous-struct/TestCallUserAnonTypedef.py
  lldb-api :: commands/expression/argument_passing_restrictions/TestArgumentPassingRestrictions.py
  lldb-api :: commands/expression/call-function/TestCallStdStringFunction.py
  lldb-api :: commands/expression/call-function/TestCallStopAndContinue.py
  lldb-api :: commands/expression/call-function/TestCallUserDefinedFunction.py
  lldb-api :: commands/expression/char/TestExprsChar.py
  lldb-api :: commands/expression/class_template_specialization_empty_pack/TestClassTemplateSpecializationParametersHandling.py
  lldb-api :: commands/expression/codegen-crash-typedefdecl-not-in_declcontext/TestCodegenCrashTypedefDeclNotInDeclContext.py
  lldb-api :: commands/expression/context-object/TestContextObject.py
  lldb-api :: commands/expression/deleting-implicit-copy-constructor/TestDeletingImplicitCopyConstructor.py
  lldb-api :: commands/expression/diagnostics/TestExprDiagnostics.py
  lldb-api :: commands/expression/expr_inside_lambda/TestExprInsideLambdas.py
  lldb-api :: commands/expression/formatters/TestFormatters.py
  lldb-api :: commands/expression/import_base_class_when_class_has_derived_member/TestImportBaseClassWhenClassHasDerivedMember.py
  lldb-api :: commands/expression/inline-namespace/TestInlineNamespace.py
  lldb-api :: commands/expression/namespace-alias/TestInlineNamespaceAlias.py
  lldb-api :: commands/expression/nested/TestNestedExpressions.py
  lldb-api :: commands/expression/no-deadlock/TestExprDoesntBlock.py
  lldb-api :: commands/expression/options/TestExprOptions.py
  lldb-api :: commands/expression/pr35310/TestExprsBug35310.py
  lldb-api :: commands/expression/rdar42038760/TestScalarURem.py
  lldb-api :: commands/expression/rdar44436068/Test128BitsInteger.py
  lldb-api :: commands/expression/static-initializers/TestStaticInitializers.py
  lldb-api :: commands/expression/test/TestExprs.py
  lldb-api :: commands/expression/unwind_expression/TestUnwindExpression.py
  lldb-api :: commands/expression/xvalue/TestXValuePrinting.py
  lldb-api :: commands/frame/recognizer/TestFrameRecognizer.py
  lldb-api :: commands/frame/var-dil/basics/NoDebugInfo/TestFrameVarDILNoDebugInfo.py
  lldb-api :: commands/platform/launchgdbserver/TestPlatformLaunchGDBServer.py
  lldb-api :: functionalities/breakpoint/breakpoint_by_line_and_column/TestBreakpointByLineAndColumn.py
  lldb-api :: functionalities/breakpoint/breakpoint_ids/TestBreakpointIDs.py
  lldb-api :: functionalities/breakpoint/breakpoint_language/TestBreakpointLanguage.py
  lldb-api :: functionalities/breakpoint/breakpoint_on_overload/TestBreakOnOverload.py
  lldb-api :: functionalities/breakpoint/breakpoint_options/TestBreakpointOptions.py
  lldb-api :: functionalities/data-formatter/data-formatter-cpp/TestDataFormatterCpp.py
  lldb-api :: functionalities/data-formatter/data-formatter-ptr-matching/TestDataFormatterPtrMatching.py
  lldb-api :: functionalities/data-formatter/data-formatter-skip-summary/TestDataFormatterSkipSummary.py
  lldb-api :: functionalities/data-formatter/data-formatter-stl/generic/forward_list/TestDataFormatterGenericForwardList.py
  lldb-api :: functionalities/data-formatter/data-formatter-stl/generic/list/TestDataFormatterGenericList.py
  lldb-api :: functionalities/data-formatter/data-formatter-stl/generic/shared_ptr/TestDataFormatterStdSharedPtr.py
  lldb-api :: functionalities/data-formatter/data-formatter-stl/generic/tuple/TestDataFormatterStdTuple.py
  lldb-api :: functionalities/data-formatter/data-formatter-stl/generic/u8string/TestDataFormatterStdU8String.py
  lldb-api :: functionalities/data-formatter/data-formatter-stl/generic/unique_ptr/TestDataFormatterStdUniquePtr.py
  lldb-api :: functionalities/data-formatter/ptr_ref_typedef/TestPtrRef2Typedef.py
  lldb-api :: functionalities/data-formatter/special-chars/TestSummaryStringSpecialChars.py
  lldb-api :: functionalities/data-formatter/stringprinter/TestStringPrinter.py
  lldb-api :: functionalities/data-formatter/typedef_array/TestTypedefArray.py
  lldb-api :: functionalities/data-formatter/var-in-aggregate-misuse/TestVarInAggregateMisuse.py
  lldb-api :: functionalities/data-formatter/varscript_formatting/TestDataFormatterVarScriptFormatting.py
  lldb-api :: functionalities/data-formatter/vector-types/TestVectorTypesFormatting.py
  lldb-api :: functionalities/inline-stepping/TestInlineStepping.py
  lldb-api :: functionalities/location-list-lookup/TestLocationListLookup.py
  lldb-api :: functionalities/memory/find/TestMemoryFind.py
  lldb-api :: functionalities/multiple-slides/TestMultipleSlides.py
  lldb-api :: functionalities/optimized_code/TestNoASanExceptionAfterEvalOP_piece.py
  lldb-api :: functionalities/return-value/TestReturnValue.py
  lldb-api :: functionalities/target_var/TestTargetVar.py
  lldb-api :: functionalities/type_find_first/TestFindFirstType.py
  lldb-api :: functionalities/type_get_module/TestTypeGetModule.py
  lldb-api :: functionalities/type_types/TestFindTypes.py
  lldb-api :: functionalities/unused-inlined-parameters/TestUnusedInlinedParameters.py
  lldb-api :: lang/c/enum_types/TestEnumTypes.py
  lldb-api :: lang/c/fpeval/TestFPEval.py
  lldb-api :: lang/c/inlines/TestRedefinitionsInInlines.py
  lldb-api :: lang/c/local_variables/TestLocalVariables.py
  lldb-api :: lang/c/set_values/TestSetValues.py
  lldb-api :: lang/c/shared_lib/TestSharedLib.py
  lldb-api :: lang/c/struct_types/TestStructTypes.py
  lldb-api :: lang/c/unicode/TestUnicodeSymbols.py
  lldb-api :: lang/cpp/break-on-initializers/TestBreakOnCPP11Initializers.py
  lldb-api :: lang/cpp/breakpoint-commands/TestCPPBreakpointCommands.py
  lldb-api :: lang/cpp/call-function/TestCallCPPFunction.py
  lldb-api :: lang/cpp/chained-calls/TestCppChainedCalls.py
  lldb-api :: lang/cpp/class-template-parameter-pack/TestClassTemplateParameterPack.py
  lldb-api :: lang/cpp/class-template-parameter-pack/TestTemplatePackArgs.py
  lldb-api :: lang/cpp/class_static/TestStaticVariables.py
  lldb-api :: lang/cpp/const_static_integral_member/TestConstStaticIntegralMember.py
  lldb-api :: lang/cpp/const_static_integral_member_int128/TestConstStaticIntegralMemberInt128.py
  lldb-api :: lang/cpp/constructors/TestCppConstructors.py
  lldb-api :: lang/cpp/dereferencing_references/TestCPPDereferencingReferences.py
  lldb-api :: lang/cpp/enum_promotion/TestCPPEnumPromotion.py
  lldb-api :: lang/cpp/extern_c/TestExternCSymbols.py
  lldb-api :: lang/cpp/forward/TestCPPForwardDeclaration.py
  lldb-api :: lang/cpp/function-qualifiers/TestCppFunctionQualifiers.py
  lldb-api :: lang/cpp/function-ref-qualifiers/TestCppFunctionRefQualifiers.py
  lldb-api :: lang/cpp/global_operators/TestCppGlobalOperators.py
  lldb-api :: lang/cpp/global_variables/TestCPPGlobalVariables.py
  lldb-api :: lang/cpp/incompatible-class-templates/TestCppIncompatibleClassTemplates.py
  lldb-api :: lang/cpp/llvm-style/TestLLVMStyle.py
  lldb-api :: lang/cpp/multiple-inheritance/TestCppMultipleInheritance.py
  lldb-api :: lang/cpp/namespace/TestNamespace.py
  lldb-api :: lang/cpp/namespace/TestNamespaceLookup.py
  lldb-api :: lang/cpp/namespace_conflicts/TestNamespaceConflicts.py
  lldb-api :: lang/cpp/nested-template/TestNestedTemplate.py
  lldb-api :: lang/cpp/nsimport/TestCppNsImport.py
  lldb-api :: lang/cpp/operators/TestCppOperators.py
  lldb-api :: lang/cpp/overloaded-functions/TestOverloadedFunctions.py
  lldb-api :: lang/cpp/preferred_name/TestPreferredName.py
  lldb-api :: lang/cpp/rvalue-references/TestRvalueReferences.py
  lldb-api :: lang/cpp/static_methods/TestCPPStaticMethods.py
  lldb-api :: lang/cpp/step-through-trampoline/TestStepThroughTrampoline.py
  lldb-api :: lang/cpp/stl/TestSTL.py
  lldb-api :: lang/cpp/structured-binding/TestStructuredBinding.py
  lldb-api :: lang/cpp/template-specialization-type/TestTemplateSpecializationType.py
  lldb-api :: lang/cpp/template/TestTemplateArgs.py
  lldb-api :: lang/cpp/type_lookup/TestCppTypeLookup.py
  lldb-api :: lang/cpp/type_lookup_duplicate/TestCppTypeLookupDuplicate.py
  lldb-api :: lang/cpp/typedef-to-outer-fwd/TestTypedefToOuterFwd.py
  lldb-api :: lang/cpp/typedef/TestCppTypedef.py
  lldb-api :: lang/cpp/unique-types2/TestUniqueTypes2.py
  lldb-api :: lang/cpp/unsigned_types/TestUnsignedTypes.py
  lldb-api :: lang/mixed/TestMixedLanguages.py
  lldb-api :: macosx/duplicate-archive-members/TestDuplicateMembers.py
  lldb-api :: python_api/frame/inlines/TestInlinedFrame.py
  lldb-api :: python_api/sbmodule/FindTypes/TestSBModuleFindTypes.py
  lldb-api :: python_api/target/TestTargetAPI.py
  lldb-api :: python_api/thread/TestThreadAPI.py
  lldb-api :: python_api/type/TestTypeList.py

********************
Unexpectedly Passed Tests (2):
  lldb-api :: lang/cpp/function_refs/TestFunctionRefs.py
  lldb-api :: lang/cpp/reference-to-outer-type/TestCppReferenceToOuterClass.py


Testing Time: 801.33s

Total Discovered Tests: 1288
  Unsupported        : 670 (52.02%)
  Passed             : 482 (37.42%)
  Expectedly Failed  :  10 (0.78%)
  Unresolved         :   5 (0.39%)
  Failed             : 119 (9.24%)
  Unexpectedly Passed:   2 (0.16%)
```

> Draft, because there are about 120 failed tests I need to check.

So if you are on Windows X64, then it's a good sign we got similar results on Windows on Arm. A couple of those are random timeouts unrelated to PDB, so it's not exactly the same.

Unfortunately I don't have time to go digging into them myself, so I am thinking about a sustainable way forward for what you want to do.

One idea is to opt-in tests to being run with PDB. This means you can do that for your formatter changes, which keeps reviewers happy. The other stuff we open an issue for the problem that forcing PDB breaks a bunch of existing tests.

I like what you're doing with the formatters and don't want to derail you from that.

DavidSpickett commented:
Opened https://github.com/llvm/llvm-project/issues/149498 so you can refer to it.

I think it's a reasonable pitch to say:
* Formatters that work with PDB are a good idea.
* If we opt-in those formatters for testing using PDB, they will be run on at least Linaro's Windows on Arm bot, maybe the x86 one too.
* One day we might clean up the rest, and by we I mean future people who may or may not be us :)

Nerixyz commented:
I made it opt-in now. A test can set `TEST_WITH_PDB_DEBUG_INFO = True`. Unfortunately, all STL tests currently fail. Most of this is because the demangled name in the native PDB parser has spaces at different positions than the DIA parser which I used before.

I'm not sure if it's possible to test both the native parser _and_ the DIA one, since they're selected by an environment variable.

Michael137 commented:
I think as a pre-requisite to this we should remove the non-native PDB parser (and anything associated with it). That way we wouldn't need any special environment variable setting etc.

We agreed that the native parser would be the way forward in the last EuroLLVM round-table IIRC (CC @JDevlieghere @labath). I think Jonas had a PR open for it but there were some cases small test blockers. What was the outcome of that work @JDevlieghere ?

DavidSpickett commented:
https://github.com/llvm/llvm-project/issues/114906 as well.

Nerixyz committed:
[LLDB] Run API tests with PDB too

Michael137 reviewed (commented):
SGTM

I'll let @labath / @JDevlieghere give the final say

Nerixyz commented on lldb/packages/Python/lldbsuite/test/lldbtest.py:
```diff
@@ -804,6 +804,13 @@ def setUpCommands(cls):
         return commands
 
+    def getDebugInfoSetupCommands(self):
+        if self.getDebugInfo() == "native-pdb":
```
`setUpCommands` is a `@classmethod`, so it gets the calling class type as it's argument, not the instance. I don't know if/why `@classmethod` is needed there.

labath replied:
I don't know that for sure, but I can imagine it being added just because the function didn't need the instance argument at the time. It may be possible to remove it.

That said, I'm not entirely happy with how these new categories are are behaving quite differently from the other debug info categories. The other categories change the kind of debug info that's being emitted -- not which code is parsing that debug info.

Having a `pdb` category to get clang to produce pdb debug info makes perfect sense. I'm not so sure about using that category to select the plugin kind -- particularly if the end goal is to remove the non-native plugin completely.

When I made the suggestion to use settings I was imagining you would be setting it on a per-run basis via the `--setting` flag to dotest.py. Would that be an option?

Nerixyz replied:
Yea, my goal is to run some tests with both PDB plugins in CI (e.g. to check that STL formatters work with both plugins + DWARF).

JDevlieghere replied:
I share Pavel's concern. To give a bit more context, we have a paragraph that talks about test variants on the website: https://lldb.llvm.org/resources/test.html#api-tests

The TL;DR is that there's many interesting variants but that leads to combinatorial explosion. Maybe testing both would be better suited for a bot configuration, like the [Matrix Bot](https://green.lab.llvm.org/job/llvm.org/view/LLDB/job/lldb-cmake-matrix/) on GreenDragon?

DavidSpickett replied:
As one half of Windows lldb bots, I can say that Linaro is not going to put another builder online for this. Pretty safe to assume that @vvereschaka would not want to either.

> Having a pdb category to get clang to produce pdb debug info makes perfect sense. I'm not so sure about using that category to select the plugin kind -- particularly if the end goal is to remove the non-native plugin completely.

Given this end goal. Once we are down to one plugin, then PDB can act as a debug info category like the others.

So in the meantime, I would be happy to do is add an extra `check-whatever` to our Windows on Arm bot.

> When I made the suggestion to use settings I was imagining you would be setting it on a per-run basis via the --setting flag to dotest.py. Would that be an option?

We could implement this, ideally with a reduced scope `check-lldb-pdbstuff`. 

Or we could use dotest.py directly? Anything that cuts down the time because the test step is already around 25 minutes.

It's ugly but could you come up with `dotest.py -p <giant regex that covers all the things we think we need>`?

DavidSpickett replied:
> We could implement this, ideally with a reduced scope check-lldb-pdbstuff.
> 
> Or we could use dotest.py directly? Anything that cuts down the time because the test step is already around 25 minutes.
> 
> It's ugly but could you come up with dotest.py -p <giant regex that covers all the things we think we need>?

Actually, whatever way this is done, a `check-lldb-supersecretpdbthing` is better because then I only have to add that to llvm-zorg once. Any future changes to what gets run can be done in llvm-project and take effect right away.

Michael137 replied:
How unusable is the Native PDB plugin to debug with? Can we just make it the default now? I know there's some tests that pass with DIA PDB that don't with the native one, but are those actually blockers?

Nerixyz committed:
fix: use single pdb debug info option

Nerixyz committed:
Merge remote-tracking branch 'upstream/main' into feat/pdb-tests

Nerixyz committed:
fix: use `-g -gcodeview`

Nerixyz commented:
Ping, now that the native plugin is the default.

I updated the branch and switched from using just `-g` to `-g -gcodeview`. `-g` ensures `-debug` is passed to the linker and `-gcodeview` tells clang to generate CodeView (and a PDB in the end) - `-gcodeview` alone doesn't work. Since it now forces codeview, this should work on targets defaulting to DWARF (MinGW) as well (@mstorsjo).

Test cases only run with PDB if `TEST_WITH_PDB_DEBUG_INFO` is set to `True` and they only use the default plugin (native).

JDevlieghere reviewed (approved):
🥳
